### ISSUE TITLE
A CookieJar cannot be shared between threads, even with a Mutex

### ISSUE THREAD

**Issue by yberreby** [alexcrichton/cookie-rs#35]
``` rust
use std::sync::{Arc, Mutex};
use std::thread;

extern crate cookie;
use cookie::{Cookie, CookieJar};

fn main() {
    let m = Arc::new(Mutex::new(CookieJar::new(b"f8f9eaf1ecdedff5e5b749c58115441e")));

    thread::spawn(move || {
      m
    });
}
```

This fails to compile with the following error message:

```
src/main.rs:10:5: 10:18 error: the trait `core::marker::Sync` is not implemented for the type `core::cell::UnsafeCell<std::collections::hash::map::HashMap<collections::string::String, cookie::Cookie>>` [E0277]
src/main.rs:10     thread::spawn(move || {
                   ^~~~~~~~~~~~~
src/main.rs:10:5: 10:18 help: run `rustc --explain E0277` to see a detailed explanation
src/main.rs:10:5: 10:18 note: `core::cell::UnsafeCell<std::collections::hash::map::HashMap<collections::string::String, cookie::Cookie>>` cannot be shared between threads safely
src/main.rs:10     thread::spawn(move || {
                   ^~~~~~~~~~~~~
src/main.rs:10:5: 10:18 error: the trait `core::marker::Sync` is not implemented for the type `core::cell::UnsafeCell<std::collections::hash::set::HashSet<collections::string::String>>` [E0277]
src/main.rs:10     thread::spawn(move || {
                   ^~~~~~~~~~~~~
src/main.rs:10:5: 10:18 help: run `rustc --explain E0277` to see a detailed explanation
src/main.rs:10:5: 10:18 note: `core::cell::UnsafeCell<std::collections::hash::set::HashSet<collections::string::String>>` cannot be shared between threads safely
src/main.rs:10     thread::spawn(move || {
                   ^~~~~~~~~~~~~
src/main.rs:10:5: 10:18 error: the trait `core::marker::Sync` is not implemented for the type `core::cell::UnsafeCell<usize>` [E0277]
src/main.rs:10     thread::spawn(move || {
                   ^~~~~~~~~~~~~
src/main.rs:10:5: 10:18 help: run `rustc --explain E0277` to see a detailed explanation
src/main.rs:10:5: 10:18 note: `core::cell::UnsafeCell<usize>` cannot be shared between threads safely
src/main.rs:10     thread::spawn(move || {
```

I've been looking for a consumer-side solution for over an hour now, but couldn't find one. Is the problem with `CookieJar` itself?

**Comment by alexcrichton** [comment]
While unintentional, this compiler is actually correct about this. While the error message isn't so great (I think there are a few issues open on this), the underlying problem is:

1. `CookieJar<'a>` may hold `&'a CookieJar<'a>` if it's a "child" cookie jar (e.g. encrypted/signed)
2. `CookieJar` may also hold `RefCell<T>` where `T` can be one of a number of maps.
3. This means that `CookieJar<'a>` may contain `&'a RefCell<T>`
4. For `&T` to be `Send` (as `Mutex` requires) then `T` must be `Sync` (as it can be shared among multiple threads). `RefCell<T>`, however, is never `Sync`.

Basically what this means is that the cookie jar is indeed not threadsafe! (only a "root cookie jar" is). For now this means that this design just isn't compatible with sharing a jar across threads, it'd probably need to have a type-level distinction between root and child cookie jars to do so.

**Comment by filsmick** [comment]
@alexcrichton Thank you for the explanation. Since I'm only using a root CookieJar, using `unsafe impl` should therefore not lead to memory safety, right?

**Comment by alexcrichton** [comment]
If you're guaranteed to never use a child cookie jar then it should be fine, yeah.
